flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/RowDataToProtoSerializer.java

```diff
@@ -0,0 +1,256 @@
+package com.google.cloud.flink.bigquery.sink.serializer;
+
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LocalZonedTimestampType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+import org.apache.flink.table.types.logical.TimeType;
+import org.apache.flink.table.types.logical.TimestampType;
+
+import com.google.api.client.util.Preconditions;
+import com.google.cloud.bigquery.storage.v1.BigDecimalByteStringEncoder;
+import com.google.cloud.flink.bigquery.sink.exceptions.BigQuerySerializationException;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.DynamicMessage;
+import org.apache.avro.generic.GenericRecord;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** Serializer for converting Flink's {@link RowData} to BigQuery proto. */
+public class RowDataToProtoSerializer extends BigQueryProtoSerializer<RowData> {
+
+    private Descriptor descriptor;
+    private LogicalType type;
+
+    public RowDataToProtoSerializer() {}
+
+    /**
+     * Prepares the serializer before its serialize method can be called. It allows contextual
+     * preprocessing after constructor and before serialize. The Sink will internally call this
+     * method when initializing itself.
+     *
+     * @param bigQuerySchemaProvider {@link BigQuerySchemaProvider} for the destination table.
+     */
+    @Override
+    public void init(BigQuerySchemaProvider bigQuerySchemaProvider) {
+        Preconditions.checkNotNull(
+                bigQuerySchemaProvider,
+                "BigQuerySchemaProvider not found while initializing AvroToProtoSerializer");
+        Descriptor derivedDescriptor = bigQuerySchemaProvider.getDescriptor();
+        Preconditions.checkNotNull(
+                derivedDescriptor, "Destination BigQuery table's Proto Schema could not be found.");
+        this.descriptor = derivedDescriptor;
+    }
+
+    public void setLogicalType(LogicalType type) {
+        this.type = type;
+    }
+
+    @Override
+    public ByteString serialize(RowData record) throws BigQuerySerializationException {
+        try {
+            return getDynamicMessageFromRowData(record, this.descriptor, this.type).toByteString();
+        } catch (Exception e) {
```

Comment:
Can you handle more specific exceptions instead of the base Exception?

Reply:
The overall connector code expects all the exceptions thrown during serialization to be `BigQuerySerializationException`. In case any Exception goes uncaught during this process, it can propagate and disrupt with the connector operation.

Reply:
Is there any exception which needs to be addressed in the serializer before throwing it as a BigQuerySerializationException?

Reply:
No. We just need to wrap all exceptions (runtime or otherwise) originating from the serializer as BigQuerySerializationException so that the connector's error handling (downstream) can handle them appropriately.

Reply:
My question is slightly different, I am not suggesting that we let the individual exceptions be propagated to the downstream. Is there there a subset of exceptions which can be retried before they are re-thrown back to the caller?

Reply:
The exceptions originate when the obtained value does not match with the expected schema type and its constraints. I do not think retrying would help as the conditions would anyways remain the same.

Reply:
okay, makes sense.